route/src/route_generator_worker.cpp

```diff
@@ -438,5 +442,77 @@ namespace route {
         return true;
     }
 
+bool RouteGeneratorWorker::crossTrackErrorCheck(const geometry_msgs::PoseStampedConstPtr& msg, lanelet::ConstLanelet current)
+{
+  if (!world_model_->getMap()) 
+  {
+    ROS_ERROR_STREAM("WMListener received a route before a map was available. Dropping route message.");
+    return true;
+  }
+
+  auto route = world_model_->getRoute()->shortestPath();
+
+  bool out_of_llt_bounds = false;
+
+    lanelet::BasicPoint2d position;
+
+    position.x()= msg->pose.position.x;
+    position.y()= msg->pose.position.y;  
+
+    if (boost::geometry::distance(position, current.polygon2d()) > cross_track_dist)
+        {
+            cte_count++;
+
+            if(cte_count > 4)
+                return true;
+        }
+
+  if(boost::geometry::within(position, current.polygon2d()) == false) //Determine whether or not the vehicle is in the lanelet polygon
+      {
+        out_of_llt_bounds = true;
+      }
+
+    auto following_llts = world_model_->getMapRoutingGraph()->following(current);
+    lanelet::ConstLanelets filtered_llts;//TODO: Filter lanelets from following lanelets that are not a part of the route
+ 
+
+    for(auto i:following_llts)
+    {
+       for(auto j: route)
+       {
+           if (i.id() == j.id())
+            filtered_llts.push_back(i);
+       }
+    }
+
+    for(auto i: filtered_llts)
+    {
+        if (boost::geometry::within(position, i.polygon2d()) == false) 
+           out_count++;
+           
+        if (boost::geometry::distance(position, current.polygon2d()) > ll_downtrack_distance_)
+            out_of_llt_bounds = true;
+    }
+
+    if(out_count == filtered_llts.size()) //If our current position is not within the bounds of any of the following lanelets, then return true
+        out_of_llt_bounds = true;
```

Comment:
Adding to the above comment, if we only have 1 next lanelet, we just need to check the downtrack distance to that. If the condition is met, update the counter and mark false until certain number is reached like you are doing above. 
Please use downtrack distance:
https://github.com/usdot-fhwa-stol/carma-platform/blob/582abd5a6a50878bb33266bd2a7bbab568dbd233/route/src/route_generator_worker.cpp#L345